include/openPMD/auxiliary/JSON_internal.hpp

```diff
@@ -0,0 +1,235 @@
+         *        gradually filled by applying each operator[]() call also to
+         *        it.
+         *        Shared pointer shared between all instances returned by
+         *        operator[]() in order to avoid use-after-free situations.
+         *
+         */
+        std::shared_ptr< nlohmann::json > m_shadow;
+        /**
+         * @brief The sub-expression within m_originalJSON corresponding with
+         *        the current instance.
+         *
+         */
+        nlohmann::json * m_positionInOriginal;
+        /**
+         * @brief The sub-expression within m_positionInOriginal corresponding
+         *        with the current instance.
+         *
+         */
+        nlohmann::json * m_positionInShadow;
+        bool m_trace = true;
+
+        void invertShadow(
+            nlohmann::json & result, nlohmann::json const & shadow ) const;
+
+        TracingJSON(
+            std::shared_ptr< nlohmann::json > originalJSON,
+            std::shared_ptr< nlohmann::json > shadow,
+            nlohmann::json * positionInOriginal,
+            nlohmann::json * positionInShadow,
+            bool trace );
+    };
+
+    template< typename Key >
+    TracingJSON TracingJSON::operator[]( Key && key )
+    {
+        nlohmann::json * newPositionInOriginal =
+            &m_positionInOriginal->operator[]( key );
+        // If accessing a leaf in the JSON tree from an object (not an array!)
+        // erase the corresponding key
+        static nlohmann::json nullvalue;
+        nlohmann::json * newPositionInShadow = &nullvalue;
+        if( m_trace && m_positionInOriginal->is_object() )
+        {
+            newPositionInShadow = &m_positionInShadow->operator[]( key );
+        }
+        bool traceFurther = newPositionInOriginal->is_object();
+        return TracingJSON(
+            m_originalJSON,
+            m_shadow,
+            newPositionInOriginal,
+            newPositionInShadow,
+            traceFurther );
+    }
+
+    /**
+     * Check if options points to a file (indicated by an '@' for the first
+     * non-whitespace character).
+     * If yes, return the file content, if not just parse options directly.
+     *
+     * @param options as a parsed JSON object.
+     * @param considerFiles If yes, check if `options` refers to a file and read
+     *        from there.
+     */
+    nlohmann::json
+    parseOptions( std::string const & options, bool considerFiles );
+
+#if openPMD_HAVE_MPI
+
+    /**
+     * Parallel version of parseOptions(). MPI-collective.
+     */
+    nlohmann::json parseOptions(
+        std::string const & options, MPI_Comm comm, bool considerFiles );
+
+#endif
+
+    /**
+     * Recursively transform all keys in a JSON dataset to lower case.
+     * String values are unaffected.
+     * JSON objects at the following openPMD-defined locations are not affected:
+     * * adios2.engine.parameters
+     * * adios2.dataset.operators.<number>.parameters
```

Comment:
Avoid warning of the kind:
```
include/openPMD/auxiliary/JSON_internal.hpp:193: error: Unsupported xml/html tag <number> found (warning treated as error, aborting now)
```
```suggestion
     * * `adios2.engine.parameters`
     * * `adios2.dataset.operators.<number>.parameters`
```